### ISSUE TITLE
Query: Infer unicode-ness for literals in complex cases

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFramework#4978]
#4937 introduced logic to infer Unicode-ness for literals when comparing to column. It works well for simple cases when one side of comparison is ColumnExpression. Though there are complex cases in which we are not inferring the information correctly atm. For e.g.
1. If one side of binary expression is subquery selecting one element of a column then the other side should have same facets as the column being projected in subquery.
2. Operators like concat which return string type values should pass uniform information towards parent because the siblings of concat should have same unicode-ness as concat. (See test `Non_unicode_string_literals_is_used_for_non_unicode_column_with_concat`)

For cases like above, we can improve logic to propagate information. If it is too hard to do then we should add functions like `AsUnicode` & `AsNonUnicode`.

**Comment by smitpatel** [comment]
Also cases of SqlFunction like CharIndex (comes from String.Contains method) the argument should have same type mapping. Though That could be different for individual functions.

**Comment by ajcvickers** [comment]
Note: when designing/implementing this, consider the case for parameters passed to FromSql, etc. See #8895

**Comment by ajcvickers** [comment]
See also comments in #11921

**Comment by ajcvickers** [comment]
See also comments in the thread here: https://github.com/aspnet/EntityFrameworkCore/issues/6717#issuecomment-401255439

**Comment by roji** [comment]
Another issue which shows the need for a way to explicitly specify a store type: https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/561.

**Comment by dpsenner** [comment]
The description of this issue reads to me that users should in the future be forced to step away from the strong types defined in the database model just to write linq queries that work. Such a solution seems unfeasible to me. I hope that in the future entity framework core provides enough metadata to providers so that they can implement smart autoboxing casts that allow generating queries from strong types that work both for inserts, updates, fetches and linq queries.

If this feature involves introducing new data types that match data types in the sql standard this would be fine with me. For instance, equivalents of `OffsetDateTime` or `ZonedDateTime` should work with the postgres `timestamptz` and `tstzrange` datatypes while an equivalent data type for `Instant` should work with `timestamp` and `tsrange`. An equivalent of `Instant` must not work with `timestamptz` or `tstzrange`. Equivalents of `OffsetDateTime` or `ZonedDateTime` must not work with `timestamp` and `tsrange` but can (given that both actually represent utc timestamps). Unfortunately, the database types vary from one database to the other and providing a superset of them all is a tough task. This indicates to me that entity framework core must provide metadata to improve the type inference that a provider can do.

**Comment by roji** [comment]
@dpsenner I agree that ideally the CLR type should be enough and we should never need to force the user to provide extra details via EF Core-specific mechanisms, but unfortunately the reality is different. There are many examples where a single .NET type can be mapped to two different database types: we can't always expect there to be a one-to-one correspondence when mapping two very different type systems (.NET and SQL Server or PostgreSQL), and it's not practical to add new CLR types everywhere. Even when there's a one-to-one mapping, database types sometimes have facets which don't exist on the .NET side (i.e. unicode-ness in SQL Server).

One approach would be to try to improve the situation by adding inference, i.e. looking at the context when determining which mapping to use. For example, if I compare a column of type B to a .NET type that can map to either A (the default) or B, we could select B as the mapping. I'm a bit skeptical of this approach and actually prefer the user being more explicit, but it's a possible way forward.

I'm not sure exactly what you mean when yu write that EF Core "must provide metadata to improve the type inference".

**Comment by roji** [comment]
And another one: https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/567

PostgreSQL's default textual type is called `text` (unlimited length), and that's what we map .NET `string` to. But one can also map string to `char(x)`, and comparing a `char(x)` column to a `text` parameter fails. This last bit is indeed very strict on PostgreSQL's side, and may be related to some strong-typing strictness added on the Npgsql side, I'll think about it a little. But the general problem remains...

**Comment by roji** [comment]
See https://github.com/npgsql/npgsql/issues/2095 for a discussion on possible Npgsql changes which would mitigate this to some extent, may be interesting/relevant to other providers.

**Comment by dpsenner** [comment]
> I'm not sure exactly what you mean when yu write that EF Core "must provide metadata to improve the type inference".

@roji As you wrote earlier, npgql already implements a context to better guess the involved types. Currently I can observe that query parameters map to `DbType = Object` in several cases. For example:

```sql
create table sandbox.timestamped_record (
	timestamp_1 timestamp without time zone
);
```

```csharp
class TimestampedRecord  {
	public DateTime? Timestamp1 { get; set; }
}
void Query() {
	var npgslRangeAroundNow = new NpgsqlRange<DateTime>(DateTime.Now - TimeSpan.FromSeconds(2), DateTime.Now + TimeSpan.FromSeconds(2));
	var i = BackendStorage
		.TimestampedRecords
		.Where(t => t.Timestamp1.HasValue && npgslRangeAroundNow.Contains(t.Timestamp1.Value))
		.FirstOrDefault();
}
```

generates the query:

```sql
""SELECT t.timestamp_1
FROM sandbox.timestamped_record AS t
WHERE t.timestamp_1 IS NOT NULL AND ((@__npgslRangeAroundNow_0 @> t.timestamp_1) = TRUE)
LIMIT 1"
```

with the parameters:

```text
[Parameters=["@__npgslRangeAroundNow_0='[2018-08-06 15:11:20,2018-08-06 15:11:24]' (DbType = Object)"]
```

While the csharp code is strictly typed, somewhere along the way to the generated query information is dropped and causes the parameters to become DbType = Object while (with my limited understanding) it should infer DbType = tsrange because the timestamp_1 column is strongly typed to timestamp. If that column was a timestamptz it would have to infer the type tstzrange. If the data provided with the query carried some timezone information, a properly inferred type could cause the query generator to understand that it has to add `@__npgslRangeAroundNow_0 AT TIME ZONE 'UTC'`.

Does this make my thoughts more clear? Please don't hesitate to ask for more if anything I write is unclear, doesn't make any sense at all or is just plain stupid. ;-)

**Comment by roji** [comment]
Nothing you're saying is stupid, but it's not exactly right either :)

`DbType` is a database-independent ADO.NET enum property on `DbParameter`. If works great if you want to use some universal type like int or string, which all databases support. However, once you need to use any type that's database-specific, the `DbType` enum isn't sufficient anymore. This is why each ADO.NET provider has its own enum (`SqlDbType`, `NpgsqlDbType`...) to allow specifying database-specific types.

The Npgsql EF Core provider doesn't use `DbType` at all, because it doesn't need any sort of database portability (it works only with Npgsql/PostgreSQL anyway), so it only uses `NpgsqlDbType`. EF Core doesn't know about about `NpgsqlDbType` when logging, so all you get is `DbType=Object`.

To summarize, Npgsql *always* sends an explicit database type ("type OID" in PostgreSQL parlance) with every parameter it sends. The type OID sent can be controlled via `DbType`, `NpgsqlDbType`, `DataTypeName`, or you can just let Npgsql use its default by not specifying any of those. But the current implementation will always tell PostgreSQL exactly what the data type is, thus preventing inference/coercion.

You can read about `DbType` and `NpgsqlDbType` in the Npgsql [basic usage docs](http://www.npgsql.org/doc/basic-usage.html).

Note that it's still possible for *EF Core* to do better inference, i.e. select a better mapping based on the query context, thus making Npgsql send a different (and more correct) data type OID.

**Comment by dpsenner** [comment]
Thanks @roji for your valuable insights. To me `DbType=Object` was an indicator that there's something fishy going on. I'm glad that I was wrong there. Keep up your good work!

**Comment by smuthuvels** [comment]
I am using EF Core 2.0.3 and I am getting the same [issue()](https://github.com/aspnet/EntityFrameworkCore/issues/8370) for Contains search. Do you have any fix for that?

**Comment by bricelam** [comment]
I added a mechanism for query translators to provide explicit type mappings for function arguments and results in PR #13648. This allows us to fix individual cases.

**Comment by roji** [comment]
Another possible need for this: setting facets in order to allow something to participate in a set operation (currently set operations over types with differing facets isn't supported, #19129).

**Comment by roji** [comment]
Note that this issue has been a bit ambiguous; we can use this issue to track explicitly forcing a type mapping on an arbitrary node, whereas #35045 is about adding an actual `CAST()` node into SQL.
